File: backend-python/services/audit/domain/audit_log.py
"""
AuditLog Domain Entity
Business logic for audit trail
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Optional, Dict, Any


@dataclass
class AuditLog:
    """AuditLog domain entity - tracks user actions"""

    id: Optional[int]
    user_id: Optional[int]
    organization_id: Optional[int]
    action: str
    resource_type: str
    resource_id: Optional[int]
    details: Optional[Dict[str, Any]]
    ip_address: Optional[str]
    user_agent: Optional[str]
    created_at: Optional[datetime] = None

    # Valid action patterns
    VALID_ACTIONS = [
        'user.create', 'user.update', 'user.delete', 'user.change_password',
        'organization.create', 'organization.update', 'organization.delete',
        'device.create', 'device.update', 'device.delete', 'device.activate',
        'content.upload', 'content.delete', 'content.assign',
        'tag.create', 'tag.update', 'tag.delete', 'tag.list',
        'auth.login', 'auth.logout', 'auth.register'
    ]

    # Valid resource types
    VALID_RESOURCE_TYPES = ['user', 'organization', 'device', 'content', 'tag', 'auth']

    def __post_init__(self):
        """Validate audit log data"""
        if not self.action or len(self.action.strip()) == 0:
            raise ValueError("Action is required")

        if not self.resource_type or len(self.resource_type.strip()) == 0:
            raise ValueError("Resource type is required")

        if self.resource_type not in self.VALID_RESOURCE_TYPES:
            raise ValueError(
                f"Resource type must be one of: {', '.join(self.VALID_RESOURCE_TYPES)}"
            )

    def is_user_action(self) -> bool:
        """Check if this is a user-related action"""
        return self.resource_type == 'user'

    def is_organization_action(self) -> bool:
        """Check if this is an organization-related action"""
        return self.resource_type == 'organization'

    def is_device_action(self) -> bool:
        """Check if this is a device-related action"""
        return self.resource_type == 'device'

    def is_tag_action(self) -> bool:
        """Check if this is a tag-related action"""
        return self.resource_type == 'tag'

    def is_system_action(self) -> bool:
        """Check if this is a system action (no user_id)"""
        return self.user_id is None

File: cms-vite/src/features/audit/types/auditLog.ts
/**
 * Audit Log Types
 *
 * LAYER 3: DOMAIN
 * Type definitions for audit logging
 */

export interface AuditLog {
  id: number;
  user_id: number | null;
  username?: string;
  organization_id: number | null;
  organization_name?: string;
  action: string;
  resource_type: string;
  resource_id: number | null;
  details: Record<string, any> | null;
  ip_address: string | null;
  user_agent: string | null;
  created_at: string;
}

export interface AuditLogListResponse {
  logs: AuditLog[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface AuditLogFilters {
  user_id?: number;
  organization_id?: number;
  action?: string;
  resource_type?: string;
  start_date?: string;
  end_date?: string;
  page?: number;
  per_page?: number;
  sort_by?: string;
  sort_dir?: 'asc' | 'desc';
}

export const RESOURCE_TYPES = [
  'user',
  'organization',
  'device',
  'content',
  'auth',
] as const;

export const ACTION_CATEGORIES = {
  user: ['user.create', 'user.update', 'user.delete', 'user.change_password'],
  organization: ['organization.create', 'organization.update', 'organization.delete'],
  device: ['device.create', 'device.update', 'device.delete', 'device.activate'],
  content: ['content.upload', 'content.delete', 'content.assign'],
  auth: ['auth.login', 'auth.logout', 'auth.register'],
} as const;
